components/builder-api/src/server/resources/channels.rs

```diff
@@ -31,354 +33,314 @@ impl Channels {
-        XHandler::new(show_package).before(opt.clone()),
-        "channel_package_latest",
-    );
-    r.get(
-        "/channels/:origin/:channel/pkgs/:pkg/:version",
-        XHandler::new(list_packages).before(opt.clone()),
-        "channel_packages_version",
-    );
-    r.get(
-        "/channels/:origin/:channel/pkgs/:pkg/:version/latest",
-        XHandler::new(show_package).before(opt.clone()),
-        "channel_packages_version_latest",
-    );
-    r.get(
-        "/channels/:origin/:channel/pkgs/:pkg/:version/:release",
-        XHandler::new(show_package).before(opt.clone()),
-        "channel_package_release",
-    );
-    r.put(
-        "/channels/:origin/:channel/pkgs/:pkg/:version/:release/promote",
-        XHandler::new(promote_package).before(basic.clone()),
-        "channel_package_promote",
-    );
-    r.put(
-        "/channels/:origin/:channel/pkgs/:pkg/:version/:release/demote",
-        XHandler::new(demote_package).before(basic.clone()),
-        "channel_package_demote",
-    );
-    r.post(
-        "/channels/:origin/:channel",
-        XHandler::new(create_channel).before(basic.clone()),
-        "channel_create",
-    );
-    r.delete(
-        "/channels/:origin/:channel",
-        XHandler::new(delete_channel).before(basic.clone()),
-        "channel_delete",
-    );
-*/
-
-/*
-
-fn list_channels(req: &mut Request) -> IronResult<Response> {
-    let origin_name = match get_param(req, "origin") {
-        Some(origin) => origin,
-        None => return Ok(Response::with(status::BadRequest)),
-    };
-
-    let mut request = OriginChannelListRequest::new();
-    request.set_include_sandbox_channels(false);
-
-    match helpers::get_origin(req, &origin_name) {
-        Ok(origin) => request.set_origin_id(origin.get_id()),
-        Err(err) => return Ok(render_net_error(&err)),
-    }
+        let mut request = OriginPackageGet::new();
+        request.set_ident(ident.clone());
+        request.set_visibilities(helpers::all_visibilities());
 
-    // Pass ?sandbox=true to this endpoint to include sanbox channels in the list. They are not
-    // there by default.
-    if let Some(sandbox) = helpers::extract_query_value("sandbox", req) {
-        if sandbox == "true" {
-            request.set_include_sandbox_channels(true);
+        let package = match route_message::<OriginPackageGet, OriginPackage>(req, &request) {
+            Ok(p) => p,
+            Err(e) => {
+                error!("Error retrieving package {:?}", &e);
+                return Err(e.into());
+            }
+        };
+
+        let mut promote = OriginPackagePromote::new();
+        promote.set_channel_id(origin_channel.get_id());
+        promote.set_package_id(package.get_id());
+        promote.set_ident(ident.clone());
+
+        match route_message::<OriginPackagePromote, NetOk>(req, &promote) {
+            Ok(_) => match route_message::<OriginPackagePromote, NetOk>(req, &promote) {
```

Comment:
Why are there two OriginPackagePromote calls?

Reply:
I'm not sure on that, this is how it is in master right now.

Reply:
I think it's a copy-pasta error - we can just call promote once..